File: DSRemote/Services/NetworkService.swift
import Foundation
import Network
import UIKit

enum ConnectionType: String {
    case wifi = "Wi-Fi"
    case usb = "USB"
}

@MainActor
class NetworkService: ObservableObject {
    @Published var isConnected = false
    @Published var connectionStatus = "Disconnected"
    @Published var games: [GameRom] = []
    @Published var discoveredHosts: [String] = []
    @Published var connectionType: ConnectionType = .wifi

    private var connection: NWConnection?
    private var host: NWEndpoint.Host = "127.0.0.1"
    private var port: NWEndpoint.Port = 9876
    private var onScreenshot: ((Data) -> Void)?
    private var onDisconnect: (() -> Void)?
    private var onAutoConnect: (() -> Void)?

    private var udpListener: NWListener?

    private var receivedBuffer = Data()

    func configure(host: String, port: Int, onScreenshot: @escaping (Data) -> Void, onDisconnect: @escaping () -> Void) {
        self.host = NWEndpoint.Host(host)
        self.port = NWEndpoint.Port(rawValue: UInt16(port)) ?? 9876
        self.onScreenshot = onScreenshot
        self.onDisconnect = onDisconnect
    }

    func setAutoConnectHandler(_ handler: @escaping () -> Void) {
        onAutoConnect = handler
    }

    var onTopScreenshot: ((Data) -> Void)?
    var onBottomScreenshot: ((Data) -> Void)?

    func startDiscovery() {
        guard let listener = try? NWListener(using: .udp, on: 9877) else { return }
        udpListener = listener

        listener.newConnectionHandler = { conn in
            conn.receiveMessage { data, _, _, _ in
                if let data = data,
                   let msg = String(data: data, encoding: .utf8),
                   msg.hasPrefix("DSREMOTE:") {
                    let parts = msg.dropFirst(9).split(separator: ":")
                    if parts.count >= 2 {
                        let ip = String(parts[0])
                        let transportType: ConnectionType = parts.count >= 4 && parts[3] == "USB" ? .usb : .wifi

                        DispatchQueue.main.async {
                            if transportType == .usb {
                                // Auto-connect via USB
                                self.host = NWEndpoint.Host(ip)
                                self.connect()
                                self.onAutoConnect?()
                            } else {
                                // WiFi: show in list for manual tap
                                if !self.discoveredHosts.contains(ip) {
                                    self.discoveredHosts.append(ip)
                                }
                            }
                        }
                    }
                }
                conn.cancel()
            }
            conn.start(queue: .main)
        }
        listener.start(queue: .main)
    }

    func connect() {
        connectionStatus = "Connecting..."
        let tcpOptions = NWProtocolTCP.Options()
        tcpOptions.keepaliveCount = 3
        tcpOptions.keepaliveIdle = 5
        tcpOptions.keepaliveInterval = 5

        let params = NWParameters(tls: nil, tcp: tcpOptions)
        connection = NWConnection(host: host, port: port, using: params)

        connection?.stateUpdateHandler = { [weak self] state in
            Task { @MainActor in
                switch state {
                case .ready:
                    self?.isConnected = true
                    self?.connectionStatus = "Connected"
                    self?.receiveLoop()
                    self?.sendDeviceInfo()
                case .failed(let error):
                    self?.isConnected = false
                    self?.connectionStatus = "Failed: \(error.localizedDescription)"
                    self?.onDisconnect?()
                case .cancelled:
                    self?.isConnected = false
                    self?.connectionStatus = "Disconnected"
                    self?.onDisconnect?()
                case .waiting(let error):
                    self?.connectionStatus = "Waiting: \(error.localizedDescription)"
                default:
                    break
                }
            }
        }

        connection?.start(queue: .main)
    }

    func disconnect() {
        connection?.cancel()
        connection = nil
        isConnected = false
        connectionStatus = "Disconnected"
        connectionType = .wifi
        games = []
    }

    private func receiveLoop() {
        connection?.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, isComplete, error in
            Task { @MainActor [weak self] in
                guard let self = self else { return }

                if let data = data, !data.isEmpty {
                    self.processData(data)
                }

                if isComplete || error != nil {
                    self.isConnected = false
                    self.connectionStatus = "Disconnected"
                    self.onDisconnect?()
                    return
                }

                self.receiveLoop()
            }
        }
    }

    private func processData(_ data: Data) {
        receivedBuffer.append(data)

        let markers: [(Data, (Data) -> Void)] = [
            ("BOTTOM_IMAGE:".data(using: .utf8)!, { [weak self] in self?.onBottomScreenshot?($0) ?? self?.onScreenshot?($0) }),
            ("TOP_IMAGE:".data(using: .utf8)!, { [weak self] in self?.onTopScreenshot?($0) ?? self?.onScreenshot?($0) }),
            ("IMAGE:".data(using: .utf8)!, { [weak self] in self?.onScreenshot?($0) }),
        ]

        while true {
            var matched = false
            for (marker, handler) in markers {
                guard receivedBuffer.count >= marker.count,
                      receivedBuffer[..<marker.count] == marker else { continue }
                let rest = receivedBuffer[marker.count...]
                if let endIndex = rest.firstIndex(of: UInt8(ascii: "\n")) ?? rest.firstIndex(of: UInt8(ascii: "\0")) {
                    let base64Data = rest[..<endIndex]
                    if let base64Str = String(data: base64Data, encoding: .utf8),
                       let imgData = Data(base64Encoded: base64Str) {
                        handler(imgData)
                    }
                    receivedBuffer = Data(rest[rest.index(after: endIndex)...])
                    matched = true
                    break
                }
            }
            if !matched { break }
        }

        while true {
            guard let newlineIndex = receivedBuffer.firstIndex(of: UInt8(ascii: "\n")) else { break }
            let jsonData = receivedBuffer[..<newlineIndex]
            receivedBuffer = Data(receivedBuffer[receivedBuffer.index(after: newlineIndex)...])

            guard !jsonData.isEmpty,
                  let json = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else { continue }

            if let action = json["action"] as? String {
                if action == "disconnected" {
                    isConnected = false
                    connectionStatus = "Disconnected by server"
                    onDisconnect?()
                } else if action == "welcome", let connType = json["connection"] as? String {
                    connectionType = connType == "usb" ? .usb : .wifi
                    connectionStatus = "Connected via \(connectionType.rawValue)"
                }
            } else if let gamesData = json["games"] as? [[String: Any]] {
                parseGames(gamesData)
            } else if let message = json["message"] as? String {
                connectionStatus = message
            }
        }
    }

    private func parseGames(_ gamesData: [[String: Any]]) {
        var parsed: [GameRom] = []
        for g in gamesData {
            guard let name = (g["name"] as? String) ?? (g["Name"] as? String),
                  let path = (g["fullPath"] as? String) ?? (g["FullPath"] as? String) else { continue }
            let platform: String = {
                if let s = g["platform"] as? String ?? g["Platform"] as? String { return s }
                if let n = g["platform"] as? Int ?? g["Platform"] as? Int { return n == 0 ? "DS" : "ThreeDS" }
                return ""
            }()
            let size = (g["sizeFormatted"] as? String ?? g["SizeFormatted"] as? String) ?? ""
            parsed.append(GameRom(name: name, fullPath: path, platform: platform, sizeFormatted: size))
        }
        games = parsed
    }

    func sendInput(_ type: InputType, args: [Float]) {
        let event = InputEvent(type: type, args: args)
        guard let data = try? JSONEncoder().encode(event) else { return }
        send(data)
    }

    func sendLaunchGame(_ game: GameRom) {
        let msg = ["action": "launch", "path": game.fullPath]
        guard let data = try? JSONSerialization.data(withJSONObject: msg) else { return }
        send(data)
    }

    func sendDisconnect() {
        let msg = ["action": "disconnect"]
        guard let data = try? JSONSerialization.data(withJSONObject: msg) else { return }
        send(data)
    }

    func sendStopEmulation() {
        let msg = ["action": "stop"]
        guard let data = try? JSONSerialization.data(withJSONObject: msg) else { return }
        send(data)
    }

    private func sendDeviceInfo() {
        let info = ["action": "deviceInfo", "name": UIDevice.current.name]
        guard let data = try? JSONSerialization.data(withJSONObject: info) else { return }
        send(data)
    }

    private func send(_ data: Data) {
        connection?.send(content: data, completion: .contentProcessed { _ in })
    }
}

File: DSRemote/Services/NetworkService.cs
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using DSRemote.Models;

namespace DSRemote.Services;

public class NetworkService : IDisposable
{
    private TcpListener? _listener;
    private TcpClient? _client;
    private NetworkStream? _stream;
    private CancellationTokenSource? _cts;
    private readonly int _port;

    public event Action<ConnectionInfo>? DeviceConnected;
    public event Action? DeviceDisconnected;
    public event Action<string>? MessageReceived;
    public event Action<byte[]>? ScreenshotReceived;

    public bool IsConnected => _client?.Connected ?? false;
    public ConnectionInfo? CurrentConnection { get; private set; }
    public string EmulatorName { get; set; } = "Citra";

    public NetworkService(int port = 9876)
    {
        _port = port;
    }

    public Task StartServer()
    {
        _listener?.Stop();
        _cts?.Cancel();
        _cts = new CancellationTokenSource();
        _listener = new TcpListener(IPAddress.Any, _port);
        _listener.Start();

        _ = Task.Run(() => AcceptLoop(_cts.Token));
        return Task.CompletedTask;
    }

    private async Task AcceptLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                var client = await _listener!.AcceptTcpClientAsync(ct);
                _client = client;
                _stream = client.GetStream();

                var remoteEp = (IPEndPoint)client.Client.RemoteEndPoint!;
                var connType = IsUsbAddress(remoteEp.Address) ? ConnectionType.USB : ConnectionType.WiFi;

                var info = new ConnectionInfo
                {
                    DeviceName = $"Device_{remoteEp.Address}",
                    RemoteAddress = remoteEp.Address,
                    Port = _port,
                    IsConnected = true,
                    ConnectedAt = DateTime.Now,
                    Type = connType
                };
                CurrentConnection = info;
                DeviceConnected?.Invoke(info);

                // Send welcome with connection type and emulator name
                var welcome = JsonSerializer.Serialize(new
                {
                    action = "welcome",
                    connection = connType == ConnectionType.USB ? "usb" : "wifi",
                    emulator = EmulatorName
                });
                _ = SendMessage(welcome);

                await ReadLoop(ct);
            }
            catch (OperationCanceledException) { break; }
            catch { continue; }
        }
    }

    private static bool IsUsbAddress(IPAddress address)
    {
        var bytes = address.GetAddressBytes();
        if (bytes[0] == 169 && bytes[1] == 254) return true;
        if (bytes[0] == 192 && bytes[1] == 168 && bytes[2] == 42) return true;
        return false;
    }

    private async Task ReadLoop(CancellationToken ct)
    {
        var buffer = new byte[8192];
        try
        {
            while (!ct.IsCancellationRequested && _stream != null)
            {
                var len = await _stream.ReadAsync(buffer, ct);
                if (len == 0) break;

                var data = buffer[..len];
                ProcessData(data);
            }
        }
        catch { }
        finally
        {
            _ = Disconnect();
        }
    }

    private void ProcessData(byte[] data)
    {
        var raw = Encoding.UTF8.GetString(data);
        if (raw.StartsWith("{"))
        {
            MessageReceived?.Invoke(raw);
            return;
        }

        var imgMarker = "IMAGE:";
        if (raw.StartsWith(imgMarker))
        {
            var base64 = raw[imgMarker.Length..].TrimEnd('\0');
            try
            {
                var imgBytes = Convert.FromBase64String(base64);
                ScreenshotReceived?.Invoke(imgBytes);
            }
            catch { }
        }
    }

    public async Task SendMessage(string message)
    {
        if (_stream == null) return;
        try
        {
            var data = Encoding.UTF8.GetBytes(message + "\n");
            await _stream.WriteAsync(data);
        }
        catch { }
    }

    public async Task SendInput(InputType type, params float[] args)
    {
        var msg = JsonSerializer.Serialize(new
        {
            type = type.ToString(),
            args
        });
        await SendMessage(msg);
    }

    private static readonly JsonSerializerOptions _jsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        Converters = { new System.Text.Json.Serialization.JsonStringEnumConverter() }
    };

    public async Task SendGameList(IEnumerable<GameRom> games)
    {
        try
        {
            var msg = JsonSerializer.Serialize(new { games }, _jsonOptions);
            await SendMessage(msg);
        }
        catch { }
    }

    public async Task Disconnect()
    {
        try
        {
            await SendMessage("{\"action\":\"disconnected\"}");
        }
        catch { }
        _stream?.Close();
        _client?.Close();
        CurrentConnection = null;
        DeviceDisconnected?.Invoke();
    }

    public void Dispose()
    {
        _cts?.Cancel();
        _stream?.Dispose();
        _client?.Dispose();
        _listener?.Stop();
    }
}

public enum InputType
{
    ButtonDown,
    ButtonUp,
    TouchDown,
    TouchMove,
    TouchUp,
    JoystickMove,
    DPadPress
}
